File: nn_simulator/model/device/datasheet/factory.py
from nn_simulator.model.device.datasheet.Datasheet import Datasheet
from typing import Dict


def from_density(
        density: float,
        size: int,
        wires_length: float,
        seed: int = Datasheet.seed
) -> Datasheet:
    """
    Returns a Datasheet that represents a device with a given density.

    Parameters
    ----------
    density: float
        Desired network density
    size: int
        Device package size (micro-meter)
    wires_length: float
        Length of a single nanowire (micro-meter)
    seed: int
        Generation seed

    Returns
    -------
    The datasheet of a device respecting the specified constraints.
    """

    # calculate number of needed wires to reach the density
    wires = int(density * size ** 2 / wires_length ** 2)

    return Datasheet(
        wires_count=wires, Lx=size, Ly=size,
        mean_length=wires_length, std_length=wires_length * 0.35,
        seed=seed
    )


def from_dict(dictionary: Dict) -> Datasheet:
    """
    Returns a Datasheet from its dict representation.

    Parameters
    ----------
    dictionary: Dict
        The dictionary representation of the datasheet
    Returns
    -------
    The datasheet of a device respecting the constraints specified in the
    dictionary.
    """

    # get a default datasheet
    datasheet = Datasheet()

    # update the default configuration with the data specified in the dictionary
    datasheet.__dict__.update(dictionary)

    return datasheet

File: src/device/datasheet/factory.jl
export from_density

"""
Returns a Datasheet that describes a device with the required characteristics.

# Parameters:
- `density::Real`: Desired network density.
- `size: Real`: Device package size in µm (micro-meters).
- `wires_length: Real`: Length of a single nanowire in µm (micro-meters).
- `seed: Int`: Device generation seed.

# Returns:
- `Datasheet`: The datasheet of a device respecting the specified constraints.
"""
function from_density(
        density::Real,
        size::Real = 500,
        wires_length::Real = 14.0,
        seed::Int = 1234
)::Datasheet

    @info "Generating a datasheet with: " density

    # calculate number of wires needed to reach the density
    wires = Int(floor(density * size ^ 2 / wires_length ^ 2))

    return Datasheet(
        wires_count = wires,
        wires_length_mean = wires_length,
        wires_length_variance = wires_length * 0.35,
        size = size,
        seed = seed
    )
end
